### DISCUSSION TITLE
How to cancel or skip an ActivityTask which status is SCHEDULED

### DISCUSSION THREAD

**Discussion by maeer** [uber/cadence#4412] [Q&A] [answered]
hello , I'm trying to cancel an ActivityTask in a workflow, its status like this :
![image](https://user-images.githubusercontent.com/54922192/130567433-04db67af-2017-45e4-88ee-48391a24e4c0.png)

Why I want to cancel ActivityTask:
Sometimes, To facilitate debugging,I want to specify the output of the ActivityTask, not run by the worker. In that case, the ActivityTask which is SCHEDULED should just be skipped ,then the next ActivityTask in the workflow should be SCHEDULED , and I use WorkflowExecutionSignaled event to specify  the output of the skipped ActivityTask. And I want the ActivityTask which is skipped not polled by worker anymore.

Problem:
Now, I can not skip an ActivityTask, do you some ideas ? 
Thank you !

**Comment by longquanzheng** [comment, marked as answer]
There is no such tool to skip a scheduled activity task . But You could write the workflow in a way that on receiving a signal then skip waiting for the results of the activity. 

Could you tell more about the debugging? Do you just want to run a specific activity?

**Reply by maeer** [reply]
“receiving a signal then skip waiting for the results of the activity”, I did exactly that. But there is a problem is that the skipped ActivityTask can be polled by worker, it  waste of resources, I skipped the ActivityTask just want the worker did not run for it .
"you just want to run a specific activity?" if there is an idea to just run a specific activity will be helpful too.

**Reply by longquanzheng** [reply]
I believe there must be some misunderstanding. 

Since you defined the activity in workflow, why do you want to skip? Why not delete from the code? 

If an activity is already scheduled, it will be started really soon. If there is anything wrong with activity worker that it can’t complete, you could fix the bug and reset the workflow. this is the typical solution in Cadence

**Reply by maeer** [reply]
“If there is anything wrong with activity worker that it can’t complete, you could fix the bug and reset the workflow”，I actually recommend this too.But sometimes fix the bug may cost lots of time , and if the ActivityTask behind the wrong activity worker depend the output of the wrong one,then this ActivityTask even the workflow  can not run correctly. Sometimes we can see that there are obvious errors in the output. At this time, we hope to manually remove this error and let the workflow complete，and then fix the bug .
This is why we want to manually give the output.

This is how we skip the ActivityTask:
First, we reset to the wrong ActivityTask;
Then,we manually remove obvious errors of the output as new output;
FInally,we give the new output to the ActivityTask(did not report to cadence service, only use signal send the new output to workflow worker instead of waiting for the results of the activity. );

**Reply by longquanzheng** [reply]
ohh okay. Have you tried the complete activity CLI command? That’s probably the best way for your case. 

Another way is to change the code of next activity to be depend on either the (wrong) activity output OR a signal. Then after deploy the code you can simply send a signal to the workflow. 

If this is a common pattern for your workflow, you should make some helper for “executing activity with waiting for signal “.

**Reply by longquanzheng** [reply]
Check “cadence workflow activity -h” to see how to use the CLI to complete an activity for a workflow

**Reply by longquanzheng** [reply]
Actually this may be what you’re looking for https://stackoverflow.com/questions/65118584/solutions-to-fix-stuck-timers-activities-in-cadence-swf-stepfunctions/65118585#65118585

see the stuck activity section

**Reply by longquanzheng** [reply]
^ complete activity command probably doesn’t work, because your activity is Scheduled not Started yet. Is that the problem? 

If so I think we probably need another command to start the activity manually. We don’t have it yet. If you are interested I can point out how to implement. It should be pretty easy/straightforward and similar to the complete activity command.

**Reply by maeer** [reply]
I can not complete the activity because activity is Scheduled not Started. This is the problem! 
I would be very grateful if you can give a solution to Started the activity.

**Reply by longquanzheng** [reply]
Actually there is no way to start a specific activity. Activity is started when `PollForActivityTask` is called
https://github.com/uber/cadence-idl/blob/48a14966426238c15570a13deab5a7905c0a89c1/thrift/shared.thrift#L1188
Only domain/tasklist can be specified. 

What is the reason that your activity couldn't start? If it's because no worker polling for the domain/tasklist, you can just set up one to let it happen, right?

**Reply by maeer** [reply]
of course I can set up a worker to make the ActivityTask start, but at the same time , other ActivityTask in the same tasklist would be polled. If I want to recognize the ActivityTask should be skipped , I must record the ActivityTask that  I have skipped in the database, and then complete the ActivityTask by the output  that I specific. That's a solution.  I want to know if cadence have better solution . Because concurrency issues need to be considered when writing to the database（that will happend when someone skip same task at the same time with different output ）

**Reply by longquanzheng** [reply]
Why would you need to recognize when starting? Why not let all of them start but not complete? Basically you can write an activity implementation to do sleep all the time and never return.

**Reply by longquanzheng** [reply]
You can recognize in the activity implementation. Basically you can get workflow ID and runID within the activity, and decides if you want to return or just sleep forever.

**Reply by maeer** [reply]
Sorry, I don’t quite understand the method you proposed. Do you mean put a worker to poll all the task in the tasklist, and determine whether complete by workflow ID and runID? That also need to store the workflow ID and runID of the skip task.

**Reply by maeer** [reply]
And if there is difficult to comminute in English, we can use Chinese. I use English just want to make others who have the same question easy to understand

**Reply by longquanzheng** [reply]
Yes basically. You should be able to do what you want by implementing the activity. 

If you only want to complete activity by CLI, the activity implementation can just do a sleep forever then you can use CLI to complete it. 

And sure, we can have some adhoc discussion on Slack. Either Chinese or English is fine. You should be able to find me on Cadence Slack.

**Reply by longquanzheng** [reply]
In Cadence there is no API start an activity by workflowID + activityID. The tasks are stored by tasklist and when calling PollForActivityTask to start them, worker doesn’t know which task until it receives one  from the API response. 

Technically it’s possible to add a new API like you said, just a bit tricky. I don’t think it’s worth the effort as we should have other way around. 

For example, let a worker run to start all the activities but not return. Then you should be able to do whatever you want. There is no much difference compared to complete an scheduled activity, I believe.

**Reply by maeer** [reply]
OK, I'll  use this method to solve this problem. 
Thanks for your answer
